hazelcast/src/main/java/com/hazelcast/ringbuffer/impl/ReadResultSetImpl.java

```diff
@@ -38,36 +42,52 @@
 
 /**
  * A list for the {@link com.hazelcast.ringbuffer.impl.operations.ReadManyOperation}.
- *
+ * <p>
  * The problem with a regular list is that if you store Data objects, then on the receiving side you get
  * a list with data objects. If you hand this list out to the caller, you have a problem because he sees
  * data objects instead of deserialized objects.
+ * <p>
+ * The predicate, filter and projection may be {@code null} in which case all elements are returned
+ * and no projection is applied.
  *
- * @param <E>
+ * @param <O> deserialized ringbuffer type
+ * @param <E> result set type, is equal to {@code O} if the projection
+ *            is {@code null} or returns the same type as the parameter
  */
-public class ReadResultSetImpl<E> extends AbstractList<E>
-        implements IdentifiedDataSerializable, HazelcastInstanceAware, ReadResultSet<E> {
+public class ReadResultSetImpl<O, E> extends AbstractList<E>
+        implements IdentifiedDataSerializable, HazelcastInstanceAware, ReadResultSet<E>, Versioned {
 
+    protected transient SerializationService serializationService;
     private transient int minSize;
     private transient int maxSize;
-    private transient IFunction<Object, Boolean> filter;
-    private transient HazelcastInstance hz;
+    private transient IFunction<O, Boolean> filter;
+    private transient Predicate<? super O> predicate;
```

Comment:
None. `IFunction` is the legacy way of doing filtering. I would prefer to keep `Predicate` since it is more common, explicit and doesn't box booleans.